packages/svelte/src/internal/client/runtime.js

```diff
@@ -586,6 +586,12 @@ export function schedule_effect(signal) {
 		effect = effect.parent;
 		var flags = effect.f;
 
+		// Branch effects are not part of the reactive signal graph as they can never have
+		// dependencies. As such, we can use effects can be CLEAN or MAYBE_DIRTY to signal
+		// that they contain an effect that is dirty and needs visiting in `process_effects`,
+		// and if not we can skip the branch entirely. This also doubles as being able to
+		// skip propagation up the graph in `schedule_effect` if we encounter a branch that
+		// is already MAYBE_DIRTY.
```

Comment:
Still trying to wrap my head round this, it's very cryptic. If I understand correctly, we're saying that since branch effects have no dependencies and can therefore never be dirty, we can (ab)use the `MAYBE_DIRTY` flag on branches to stop traversing upwards to the root. Is that right?

If so, it feels suboptimal — it means we traverse upwards unnecessarily until we hit a branch, and it means we call `check_dirtiness` for branches unnecessarily. What if we had an `EFFECT_HAS_DIRTY_CHILDREN` flag or similar instead?

Reply:
>  If I understand correctly, we're saying that since branch effects have no dependencies and can therefore never be dirty, we can (ab)use the MAYBE_DIRTY flag on branches to stop traversing upwards to the root. Is that right?

Not entirely, it's mainly to avoid traversing it in `process_effects` which is where the real expense is as that's a full tree search without this optimisation in place.

> If so, it feels suboptimal — it means we traverse upwards unnecessarily until we hit a branch, and it means we call check_dirtiness for branches unnecessarily.

We hit a branch traversing up almost always 1 or 2 levels, it's actually very effective.

> What if we had an EFFECT_HAS_DIRTY_CHILDREN flag or similar instead?

That would just mean having more code in the codebase. I don't understand why this is a problem, this code has been around since we've had effect scheduling based on roots.

Reply:
> That would just mean having more code in the codebase

Would it? It seems we could probably _simplify_ `process_effects` if 'this effect is/could be dirty' and 'this effect has dirty children' were separate concepts, instead of needing to check `is_branch` at numerous points inside that function. And `schedule_effect` would get simpler, especially since we wouldn't need a chunky comment explaining what's going on.

The performance benefits might be negligible but the real win would be more self-explanatory code, I think

Reply:
I added `EFFECT_HAS_DIRTY_CHILDREN`. Feel free to refactor `process_effects` as I'm in the middle of another rabbit hole.